index.html

```diff
@@ -101,212 +220,1263 @@ <h3>
               The port that the socket should be bound to. When omitted, or 0,
               any port can be used.
             </dd>
+            <div class="note">
+              The only current implementation (Chromium) ignores
+              {{TCPSocketOptions/localAddress}} and {{TCPSocketOptions/localPort}}.
+            </div>
+            -->
+            <dt>
+              <dfn>noDelay</dfn> member
+            </dt>
+            <dd>
+              Enables the `TCP_NODELAY` option,
+              disabling <a href=
+              "https://en.wikipedia.org/wiki/Nagle%27s_algorithm">Nagle's
+              algorithm</a>.
+            </dd>
+            <dt>
+              <dfn>keepAliveDelay</dfn> member
+            </dt>
+            <dd>
+              Enables TCP Keep-Alive by setting `SO_KEEPALIVE` option on the socket equal
+              to the specified Keep-Alive ping delay in milliseconds.
+              <div class="note">
+                Keep-Alive is disabled by default.
+              </div>
+            </dd>
+          </dl>
+
+        </section>
+
+      </section>
+
+      <section>
+        <h3>{{TCPSocket/[[readable]]}} attribute (internal)</h3>
+
+        <aside class="example">
+          An application receiving data from a TCP socket will typically use the {{ReadableStream}} like this:
+
+          <pre class="js">
+            const { readable } = await socket.opened;
+
+            const reader = readable.getReader();
+
+            while (true) {
+              const { value, done } = await reader.read();
+              if (done) {
+                // |reader| has been canceled.
+                break;
+              }
+              // Do something with |value|...
+            }
+
+            reader.releaseLock();
+          </pre>
+        </aside>
+      
+        The steps to <dfn data-lt="initialize tcp readable stream">initialize the TCP readable stream</dfn> are:
+
+        <ol>
+          <li>
+            Let |stream:ReadableStream| be a [=new=] {{ReadableStream}}.
+          <li>
+            Let |pullAlgorithm| be the following steps:
+            <ol>
+              <li>
+                Run the following steps in parallel:
+              <ol>
+                <li>
+                  Invoke the operating system to read a tcp packet from the socket, placing the result in the [=byte sequence=] |bytes|.
```

Comment:
I've lost track of my original comment about this but I still think this should be changed to request up to `desiredSize` bytes from the operating system.

Reply:
Here you go: [discussion_r920567920](https://github.com/WICG/direct-sockets/pull/48#discussion_r920567920).
I believe we came to an agreement that we're reading full packets with size limited by `receiveBufferSize`.
We could also adopt the same approach as with UDP and say that every time we ask the system to provide us with up to `desiredSize` TCP packets to be stored in the internal stream queue.

Reply:
The underlying platform API doesn't allow us to read full TCP packets. When we call `Read(...)` we might get a partial packet if the buffer we provide is smaller than the amount of data in the receive buffer. The rest of the packet will be waiting for us next time we call `Read(...)`. Developers are going to write buggy code if they expect to get full TCP packets and this is not something that the TCP API really exposes anyways.